docs/design/lambdas.md

```diff
@@ -0,0 +1,555 @@
+| :----------------: | :----------------------------------------------------------------------------------------------------------------: |
+|         A1         |               Required Returned Expression ([positional parameters](#positional-parameters) allowed)               |
+|         A2         |             Required Returned Expression ([positional parameters](#positional-parameters) disallowed)              |
+|         B          |                                   [Default Capture Mode](#default-capture-mode)                                    |
+|         C          | Explicit [Captures](#function-captures), [Function Fields](#function-fields) and Deduced Parameters (in any order) |
+|         D          |                                                Explicit Parameters                                                 |
+|         E1         |           Body of Statements (no return value) ([positional parameters](#positional-parameters) allowed)           |
+|         E2         |          Body of Statements (with return value) ([positional parameters](#positional-parameters) allowed)          |
+|         E3         |         Body of Statements (no return value) ([positional parameters](#positional-parameters) disallowed)          |
+|         E4         |        Body of Statements (with return value) ([positional parameters](#positional-parameters) disallowed)         |
+|         F          |                                                Required Return Type                                                |
+|         G          |                                             Function Declaration Name                                              |
+
+```carbon
+// Lambdas (all the following are in an expression context and are
+// themselves expressions)
+
+fn => A1
+
+fn [B, C] => A1
+
+fn (D) => A2
+
+fn [B, C](D) => A2
+
+fn { E1; }
+
+fn -> F { E2; }
+
+fn [B, C] { E1; }
+
+fn [B, C] -> F { E2; }
+
+fn (D) { E3; }
+
+fn (D) -> F { E4; }
+
+fn [B, C](D) { E3; }
+
+fn [B, C](D) -> F { E4; }
+```
+
+```carbon
+// Function Declarations (all the following are allowed as statements in a
+// function body or as declarations in other scopes)
+
+fn G => A1;
+
+fn G[B, C] => A1;
+
+fn G(D) => A2;
+
+fn G[B, C](D) => A2;
+
+fn G { E1; }
+
+fn G -> F { E2; }
+
+fn G[B, C] { E1; }
+
+fn G[B, C] -> F { E2; }
+
+fn G(D) { E3; }
+
+fn G(D) -> F { E4; }
+
+fn G[B, C](D) { E3; }
+
+fn G[B, C](D) -> F { E4; }
+```
+
+## Introducer
+
+**Proposal**: Introduce with the `fn` keyword to mirror function declarations.
+If a statement or declaration begins with `fn`, a name is required and it
+becomes a function declaration. Otherwise, if in an expression context, `fn`
+introduces a lambda.
+
+```carbon
+let lambda1: auto = fn => T.Make();
+
+let lambda2: auto = fn -> T { return T.Make(); };
+
+fn FunctionDeclaration1 => T.Make();
+
+fn FunctionDeclaration2 -> T { return T.Make(); }
+```
+
+## Positional Parameters
```

Comment:
I left off here. Hopefully the above comments make it clear what changes would be good to make in the following.